### PR TITLE
Enhance lightning api, client api, LauncherExecutor

### PR DESCRIPTION
**Pull request by YuanTingHsieh**
### Description

Enhance the lightning API, modify the FLCallback from @chesterxgchen (#1906) 

- Add "global_evaluation" and "training" flag/functionality, so LauncherExecutor side know it is expecting 1 model or 1 metrics or 1 model and 1 metrics
- Add "global_evaluation" flag, so the API side knows whether to send model when training finishes and send metrics when testing finishes
- Add "wait_task" functionality in Launcher, so that LauncherExecutor will call this wait method implemented by the Launcher, this prevents LauncherExecutor to accidentally kill the unfinished train.py when LauncherExecutor receives all information it needs
- Add automatic update "config_exchange.json" functionality in LauncherExecutor, so users don't need to configure "global_evaluation"/"training" flag in multiple places
- Add a "clear" flag to Client API "send" method, so we can send multiple times without clear the cache
- Add a "clear" method in Client API, so we can clear the cache without calling send method.

### Types of changes

- [x] Non-breaking change (fix or new feature that would not break existing functionality).
- [ ] Breaking change (fix or new feature that would cause existing functionality to change).
- [ ] New tests added to cover the changes.
- [ ] Quick tests passed locally by running `./runtest.sh`.
- [ ] In-line docstrings updated.
- [ ] Documentation updated.

### CODE REVIEW

File: examples/hello-world/ml-to-fl/jobs/lightning_client_api/app/custom/autoencoder.py [hunk 1]
```diff
@@ -193,24 +193,31 @@ def predict_dataloader(self):
...
+        trainer_defaults={
+            "callbacks": [ImageSampler()],
+            "max_epochs": 1,
+        },
         save_config_kwargs={"overwrite": True},
     )
+    # (1) patch the lightning trainer
+    flare.patch(cli.trainer)
+
+    print("--- test global model ---")
```
Comment (holgerroth) [inline examples/hello-world/ml-to-fl/jobs/lightning_client_api/app/custom/autoencoder.py:211]: Let's add a comment here like `# (2) evaluate the current global model to allow server-side model selection"

File: examples/hello-world/ml-to-fl/jobs/lightning_client_api/app/custom/autoencoder.py [hunk 2]
```diff
@@ -193,24 +193,31 @@ def predict_dataloader(self):
...
+    cli.trainer.test(cli.model, datamodule=cli.datamodule)
+
+    print("--- train new model ---")
```
Comment (holgerroth) [inline examples/hello-world/ml-to-fl/jobs/lightning_client_api/app/custom/autoencoder.py:214]: `# (3) Perform local training starting with the received global model"

File: examples/hello-world/ml-to-fl/jobs/lightning_client_api/app/custom/autoencoder.py [hunk 3]
```diff
@@ -193,24 +193,31 @@ def predict_dataloader(self):
...
     cli.trainer.fit(cli.model, datamodule=cli.datamodule)
+
+    print("--- test new model ---")
```
Comment (holgerroth) [inline examples/hello-world/ml-to-fl/jobs/lightning_client_api/app/custom/autoencoder.py:217]: `# (4) optionally test the new local model`

File: nvflare/app_opt/lightning/api.py [hunk 4]
```diff
@@ -12,57 +12,96 @@
...
+        # the metrics_captured will be set to True.
+        # The subsequence test() calls will not trigger the receive update model.
+        # Hence the test() will be validating the local model.
+        if pl_module and self.has_global_eval and self.metrics is None:
+            self._receive_update_model(pl_module)
+
+    def on_test_end(self, trainer, pl_module):
+        print("YYYYYYYY calling on test end")
+        if pl_module and self.has_global_eval:
+            self.metrics = _extract_metrics(trainer.logged_metrics)
```
[the remark this answers is not in the document]
Comment (YuanTingHsieh) [reply nvflare/app_opt/lightning/api.py]: Yes, we change to "callback_metrics" according to forums: https://lightning.ai/forums/t/how-to-access-the-logged-results-such-as-losses/155 
and discussions: https://github.com/Lightning-AI/lightning/issues/13147
and ray implementation: https://github.com/Lightning-AI/lightning/issues/13147

**Comment by YuanTingHsieh** [comment]
@chesterxgchen thanks for the detailed review, changes as suggested/discussed.

@holgerroth thanks for the review, changes as suggested.

**Review by holgerroth — DISMISSED** [review]
Nice work!

File: nvflare/client/config.py [hunk 5]
```diff
@@ -52,4 +54,16 @@ def get_exchange_format(self) -> ModelExchangeFormat:
...
-        return self.config[ConfigKey.TRANSFER_TYPE]
+        return self.config.get(ConfigKey.TRANSFER_TYPE, "FULL")
+
+    def to_json(self, config_file: str):
+        with open(config_file, "w") as f:
+            json.dump(self.config, f)
+
+
+def from_file(config_file: str):
+    config = ConfigService.load_configuration(config_file)
```
[the remark this answers is not in the document]
Comment (YuanTingHsieh) [reply nvflare/client/config.py]: thanks, good suggestion, I change it to ConfigFactory.

**Pull request merged**